modmail/bot.py

```diff
@@ -19,51 +20,107 @@ class ModmailBot(commands.Bot):
 
-    async def get_prefix(self, message: discord.Message = None) -> t.List[str]:
-        """Returns the bot prefix, but also allows the bot to work with user mentions."""
-        return [self.config.bot.prefix, f"<@{self.user.id}> ", f"<@!{self.user.id}> "]
+    async def start(self, *args, **kwargs) -> None:
+        """
+        Start the bot.
+
+        This just serves to create the http session.
+        """
+        await self.create_session()
+        await super().start(*args, **kwargs)
 
     async def close(self) -> None:
-        """Safely close HTTP session and extensions when bot is shutting down."""
-        await super().close()
+        """Safely close HTTP session, unload plugins and extensions when the bot is shutting down."""
+        plugins = self.extensions & PLUGINS.keys()
+        for plug in list(plugins):
+            try:
+                self.unload_extension(plug)
+            except Exception:
+                self.logger.error(f"Exception occured while unloading plugin {plug.name}", exc_info=True)
 
         for ext in list(self.extensions):
             try:
                 self.unload_extension(ext)
             except Exception:
-                log.error(f"Exception occured while unloading {ext.name}", exc_info=1)
+                self.logger.error(f"Exception occured while unloading {ext.name}", exc_info=True)
 
         for cog in list(self.cogs):
             try:
                 self.remove_cog(cog)
             except Exception:
-                log.error(f"Exception occured while removing cog {cog.name}", exc_info=1)
-
-        await super().close()
+                self.logger.error(f"Exception occured while removing cog {cog.name}", exc_info=True)
 
         if self.http_session:
             await self.http_session.close()
 
-    def add_cog(self, cog: commands.Cog) -> None:
+        await super().close()
+
+    def load_extensions(self) -> None:
+        """Load all enabled extensions."""
+        EXTENSIONS.update(walk_extensions())
+
+        # set up no_unload global too
+        for ext, value in EXTENSIONS.items():
+            if value[1]:
+                NO_UNLOAD.append(ext)
+
+        for extension, value in EXTENSIONS.items():
+            if value[0]:
+                self.logger.debug(f"Loading extension {extension}")
+                self.load_extension(extension)
+
+    def load_plugins(self) -> None:
+        """Load all enabled plugins."""
+        PLUGINS.update(walk_plugins())
+
+        for plugin, should_load in PLUGINS.items():
+            if should_load:
+                self.logger.debug(f"Loading plugin {plugin}")
+                try:
+                    # since we're loading user generated content,
+                    # any errors here will take down the entire bot
+                    self.load_extension(plugin)
+                except Exception:
+                    self.logger.error("Failed to load plugin {0}".format(plugin), exc_info=True)
```

Comment:
In the future, we could consider splitting filling the list and loading plugins separately if we want. I've also got an itch that this'll cause nasty issues if stuff gets loaded before this into plugins (should never happen, so not worth worrying about)

Feel free to insta-resolve this, just commentary.